File: src/eca.py
#! /usr/bin/env python3

# eca.py - paint elementary cellular automata on the framebuffer

import random

RULE = int(input("Rule = "))
SEED = int(input("Seed = "))

# UNPORTABLE
BLACK = b"\x00\x00\x00\x00"
WHITE = b"\xff\xff\xff\x00"

# UNPORTABLE
XRES = 1376
YRES = 768

cells = XRES * [0]
clone = XRES * [0]

if (SEED):
  random.seed(SEED)
  for x in range(XRES):
    cells[x] = random.randint(0, 1)
else:
  cells[XRES // 2] = 1

# Disable the cursor
print("\x1b[?25l", end = "", flush = True)

with open("/dev/fb0", "wb") as fb:
  for y in range(YRES):
    for x in range(XRES):
      fb.write(WHITE if cells[x] else BLACK)

      p = cells[x - 1]
      q = cells[x]
      r = cells[(x + 1) % XRES]

      clone[x] = RULE >> (p << 2 | q << 1 | r) & 1

    cells = [*clone]

input()

# Enable the cursor and clear the screen
print("\x1b[?25h\x1b[3J", end = "")

File: src/eca.c
// eca.c - code for elementary cellular automata

#include "eca.h"

#define RAND_LSB 0
#define RAND_MSB 31

static uint32_t random = 1;

uint32_t eca_rand(void) {
  uint32_t accumulator = 0;

  for (int b = RAND_LSB; b <= RAND_MSB; ++b)
    accumulator |= (uint32_t) eca_rule30(
      (uint32_t) 1 << (b == RAND_MSB ? RAND_LSB : b + 1) & random,
      (uint32_t) 1 <<                             b      & random,
      (uint32_t) 1 << (b == RAND_LSB ? RAND_MSB : b - 1) & random) << b;

  return random = accumulator;
}

void eca_srand(uint32_t seed) {
  random = seed;
}
